// @ngx-formly/core/esm2015/lib/services/formly.form.builder.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { FormGroup, FormArray, FormControl, AbstractControl, Validators } from '@angular/forms';
import { FormlyConfig } from './formly.config';
import { FORMLY_VALIDATORS, getFieldId, isObject, isNullOrUndefined, getKeyPath, assignModelValue, isUndefined, clone, removeFieldControl, getFieldValue } from '../utils';
import * as i0 from "@angular/core";
import * as i1 from "./formly.config";
export class FormlyFormBuilder {
    /**
     * @param {?} formlyConfig
     */
    constructor(formlyConfig) {
        this.formlyConfig = formlyConfig;
        this.formId = 0;
    }
    /**
     * @param {?} formControl
     * @param {?=} fieldGroup
     * @param {?=} model
     * @param {?=} options
     * @return {?}
     */
    buildForm(formControl, fieldGroup = [], model, options) {
        /** @type {?} */
        const fieldTransforms = (options && options.fieldTransform) || this.formlyConfig.extras.fieldTransform;
        (Array.isArray(fieldTransforms) ? fieldTransforms : [fieldTransforms]).forEach(fieldTransform => {
            if (fieldTransform) {
                fieldGroup = fieldTransform(fieldGroup, model, formControl, options);
                if (!fieldGroup) {
                    throw new Error('fieldTransform must return an array of fields');
                }
            }
        });
        this._buildForm({ fieldGroup, model, formControl, options });
        if ((/** @type {?} */ (options))._checkField) {
            (/** @type {?} */ (options))._checkField({ fieldGroup, model, formControl, options });
        }
    }
    /**
     * @param {?} root
     * @return {?}
     */
    _buildForm(root) {
        this.formId++;
        root.fieldGroup.forEach((field, index) => {
            this.getExtensions().forEach(extension => extension.prePopulate && extension.prePopulate(field));
            this.initFieldOptions(root, field, index);
            this.getExtensions().forEach(extension => extension.onPopulate && extension.onPopulate(field));
            this.initFieldValidation(field);
            this.initFieldAsyncValidation(field);
            if (field.key && field.type) {
                /** @type {?} */
                const paths = getKeyPath({ key: field.key });
                /** @type {?} */
                let rootForm = /** @type {?} */ (root.formControl);
                /** @type {?} */
                let rootModel = field.fieldGroup ? { [paths[0]]: field.model } : field.model;
                paths.forEach((path, index) => {
                    /** @type {?} */
                    const formPath = path.toString();
                    // is last item
                    if (index === paths.length - 1) {
                        this.addFormControl(rootForm, field, rootModel, formPath);
                    }
                    else {
                        /** @type {?} */
                        let nestedForm = /** @type {?} */ (rootForm.get(formPath));
                        if (!nestedForm) {
                            nestedForm = new FormGroup({});
                            this.addControl(rootForm, formPath, nestedForm);
                        }
                        if (!rootModel[path]) {
                            rootModel[path] = typeof path === 'string' ? {} : [];
                        }
                        rootForm = nestedForm;
                        rootModel = rootModel[path];
                    }
                });
            }
            else if (!field.key && field.fieldGroup) {
                field.formControl = root.formControl;
            }
            if (field.fieldGroup) {
                this._buildForm(field);
            }
            this.getExtensions().forEach(extension => extension.postPopulate && extension.postPopulate(field));
        });
    }
    /**
     * @return {?}
     */
    getExtensions() {
        return Object.keys(this.formlyConfig.extensions).map(name => this.formlyConfig.extensions[name]);
    }
    /**
     * @param {?} root
     * @param {?} field
     * @param {?} index
     * @return {?}
     */
    initFieldOptions(root, field, index) {
        Object.defineProperty(field, 'options', { get: () => root.options, configurable: true });
        Object.defineProperty(field, 'parent', { get: () => root, configurable: true });
        Object.defineProperty(field, 'model', {
            get: () => field.key && field.fieldGroup ? getFieldValue(field) : root.model,
            configurable: true,
        });
        field.id = getFieldId(`formly_${this.formId}`, field, index);
        field.templateOptions = field.templateOptions || {};
        field.modelOptions = field.modelOptions || {};
        field.lifecycle = field.lifecycle || {};
        if (field.type && field.key) {
            field.templateOptions = Object.assign({
                label: '',
                placeholder: '',
                focus: false,
            }, field.templateOptions);
        }
        if (field.template && field.type !== 'formly-template') {
            if (field.type) {
                console.warn(`NgxFormly: passing 'type' property is not allowed when 'template' is set.`);
            }
            field.type = 'formly-template';
        }
        if (field.type) {
            this.formlyConfig.getMergedField(field);
        }
        if (field.key && isUndefined(field.defaultValue) && (field.fieldGroup || field.fieldArray)) {
            field.defaultValue = field.fieldArray ? [] : {};
        }
        if (!isUndefined(field.defaultValue) && isUndefined(getFieldValue(field))) {
            assignModelValue(root.model, field.key, field.defaultValue);
        }
        this.initFieldWrappers(field);
        if (field.fieldArray) {
            this.initFieldArray(field);
        }
        if (!field.type && field.fieldGroup) {
            field.type = 'formly-group';
        }
    }
    /**
     * @param {?} field
     * @return {?}
     */
    initFieldArray(field) {
        field.fieldGroup = field.fieldGroup || [];
        if (field.fieldGroup.length > field.model.length) {
            for (let i = field.fieldGroup.length; i >= field.model.length; --i) {
                removeFieldControl(/** @type {?} */ (field.formControl), i);
                field.fieldGroup.splice(i, 1);
            }
        }
        for (let i = field.fieldGroup.length; i < field.model.length; i++) {
            /** @type {?} */
            const f = Object.assign({}, clone(field.fieldArray), { key: `${i}` });
            field.fieldGroup.push(f);
        }
    }
    /**
     * @param {?} form
     * @param {?} field
     * @param {?} model
     * @param {?} path
     * @return {?}
     */
    addFormControl(form, field, model, path) {
        /** @type {?} */
        const abstractControlOptions = /** @type {?} */ ({
            validators: field._validators,
            asyncValidators: field._asyncValidators,
            updateOn: field.modelOptions.updateOn,
        });
        /** @type {?} */
        let control;
        if (field.formControl instanceof AbstractControl || form.get(path)) {
            control = field.formControl || form.get(path);
            if (!(isNullOrUndefined(control.value) && isNullOrUndefined(model[path]))
                && control.value !== model[path]
                && control instanceof FormControl) {
                control.patchValue(model[path]);
            }
            if (abstractControlOptions.validators || abstractControlOptions.asyncValidators) {
                if (abstractControlOptions.validators) {
                    control.setValidators(abstractControlOptions.validators);
                }
                if (abstractControlOptions.asyncValidators) {
                    control.setAsyncValidators(abstractControlOptions.asyncValidators);
                }
                control.updateValueAndValidity();
            }
        }
        else if ((/** @type {?} */ (field)).component && (/** @type {?} */ (field)).component.createControl) {
            control = (/** @type {?} */ (field)).component.createControl(model[path], field);
        }
        else if (field.fieldGroup && !field.fieldArray) {
            control = new FormGroup({}, abstractControlOptions);
        }
        else if (field.fieldArray) {
            control = new FormArray([], abstractControlOptions);
        }
        else {
            control = new FormControl(model[path], abstractControlOptions);
        }
        if (field.templateOptions.disabled) {
            control.disable();
        }
        // Replace decorated property with a getter that returns the observable.
        // https://github.com/angular-redux/store/blob/master/src/decorators/select.ts#L79-L85
        if (delete field.templateOptions.disabled) {
            Object.defineProperty(field.templateOptions, 'disabled', {
                get: () => !field.formControl.enabled,
                set: (value) => value ? field.formControl.disable() : field.formControl.enable(),
                enumerable: true,
                configurable: true,
            });
        }
        this.addControl(form, path, control, field);
    }
    /**
     * @param {?} form
     * @param {?} key
     * @param {?} formControl
     * @param {?=} field
     * @return {?}
     */
    addControl(form, key, formControl, field) {
        if (field) {
            field.formControl = formControl;
        }
        if (form instanceof FormArray) {
            if (form.at(/** @type {?} */ (key)) !== formControl) {
                form.setControl(/** @type {?} */ (key), formControl);
            }
        }
        else {
            if (form.get(/** @type {?} */ (key)) !== formControl) {
                form.setControl(/** @type {?} */ (key), formControl);
            }
        }
    }
    /**
     * @param {?} field
     * @return {?}
     */
    initFieldValidation(field) {
        if (field._validators) {
            return;
        }
        field._validators = [];
        this.initPredefinedFieldValidation(field);
        if (field.validators) {
            for (const validatorName in field.validators) {
                if (validatorName !== 'validation') {
                    /** @type {?} */
                    let validator = field.validators[validatorName];
                    /** @type {?} */
                    let errorPath;
                    /** @type {?} */
                    let message;
                    if (isObject(validator)) {
                        errorPath = validator.errorPath;
                        message = validator.message;
                        validator = validator.expression;
                    }
                    field._validators.push((control) => {
                        /** @type {?} */
                        const isValid = validator(control, field);
                        if (errorPath && field.formControl && field.formControl.get(errorPath)) {
                            if (!isValid) {
                                field.formControl.get(errorPath).setErrors(Object.assign({}, (field.formControl.get(errorPath).errors || {}), { [validatorName]: { message } }));
                            }
                            else {
                                /** @type {?} */
                                const errors = (field.formControl.get(errorPath).errors || {});
                                delete errors[validatorName];
                                field.formControl.get(errorPath).setErrors(Object.keys(errors).length === 0 ? null : errors);
                            }
                        }
                        return isValid ? null : { [validatorName]: errorPath ? { errorPath } : true };
                    });
                }
                else {
                    if (!Array.isArray(field.validators.validation)) {
                        field.validators.validation = [field.validators.validation];
                    }
                    field.validators.validation
                        .forEach((validator) => field._validators.push(this.wrapNgValidatorFn(field, validator)));
                }
            }
        }
    }
    /**
     * @param {?} field
     * @return {?}
     */
    initFieldAsyncValidation(field) {
        if (field._asyncValidators) {
            return;
        }
        field._asyncValidators = [];
        if (field.asyncValidators) {
            for (const validatorName in field.asyncValidators) {
                if (validatorName !== 'validation') {
                    /** @type {?} */
                    let validator = field.asyncValidators[validatorName];
                    if (isObject(validator)) {
                        validator = validator.expression;
                    }
                    field._asyncValidators.push((control) => new Promise((resolve) => {
                        return validator(control, field).then((result) => {
                            resolve(result ? null : { [validatorName]: true });
                        });
                    }));
                }
                else {
                    if (!Array.isArray(field.asyncValidators.validation)) {
                        field.asyncValidators.validation = [field.asyncValidators.validation];
                    }
                    field.asyncValidators.validation
                        .forEach((validator) => field._asyncValidators.push(/** @type {?} */ (this.wrapNgValidatorFn(field, validator))));
                }
            }
        }
    }
    /**
     * @param {?} field
     * @return {?}
     */
    initPredefinedFieldValidation(field) {
        FORMLY_VALIDATORS
            .filter(opt => field.templateOptions.hasOwnProperty(opt) || (field.expressionProperties && field.expressionProperties[`templateOptions.${opt}`]))
            .forEach((opt) => {
            field._validators.push((control) => {
                /** @type {?} */
                const value = field.templateOptions[opt];
                if (value === false) {
                    return null;
                }
                switch (opt) {
                    case 'required':
                        return Validators.required(control);
                    case 'pattern':
                        return Validators.pattern(value)(control);
                    case 'minLength':
                        return Validators.minLength(value)(control);
                    case 'maxLength':
                        return Validators.maxLength(value)(control);
                    case 'min':
                        return Validators.min(value)(control);
                    case 'max':
                        return Validators.max(value)(control);
                }
            });
        });
    }
    /**
     * @param {?} field
     * @param {?} validator
     * @return {?}
     */
    wrapNgValidatorFn(field, validator) {
        validator = typeof validator === 'string'
            ? this.formlyConfig.getValidator(validator).validation
            : validator;
        return (control) => (/** @type {?} */ (validator))(control, field);
    }
    /**
     * @param {?} field
     * @return {?}
     */
    initFieldWrappers(field) {
        field.wrappers = field.wrappers || [];
        /** @type {?} */
        const fieldTemplateManipulators = Object.assign({ preWrapper: [], postWrapper: [] }, (field.templateOptions.templateManipulators || {}));
        field.wrappers = [
            ...this.formlyConfig.templateManipulators.preWrapper.map(m => m(field)),
            ...fieldTemplateManipulators.preWrapper.map(m => m(field)),
            ...field.wrappers,
            ...this.formlyConfig.templateManipulators.postWrapper.map(m => m(field)),
            ...fieldTemplateManipulators.postWrapper.map(m => m(field)),
        ].filter((el, i, a) => el && i === a.indexOf(el));
    }
}
FormlyFormBuilder.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] },
];
/** @nocollapse */
FormlyFormBuilder.ctorParameters = () => [
    { type: FormlyConfig }
];
/** @nocollapse */ FormlyFormBuilder.ngInjectableDef = i0.defineInjectable({ factory: function FormlyFormBuilder_Factory() { return new FormlyFormBuilder(i0.inject(i1.FormlyConfig)); }, token: FormlyFormBuilder, providedIn: "root" });
if (false) {
    /** @type {?} */
    FormlyFormBuilder.prototype.formId;
    /** @type {?} */
    FormlyFormBuilder.prototype.formlyConfig;
}
